File: csharp-hfsframework/hfsframework-admin/hfsframework.admin/AdmMenuBC.cs
using System.Collections.Generic;
using hfsframework.hfsbase;
using hfsframework.admin.model;
using hfsframework.admin.data;
using hfsframework.security.model;

namespace hfsframework.admin
{
    public class AdmMenuBC: BaseBusinessController<AdmMenu, long, AdmMenuRepository>
    {
        private static AdmMenuBC instancia = new AdmMenuBC();

        public static AdmMenuBC Instancia
        {
            get
            {
                return instancia;
            }
        }

        private static List<AdmMenu> cacheMenuEstaticos = new List<AdmMenu>();

        private AdmMenuBC():base(new AdmMenuRepository())
        {
        }

        public AdmMenu salvarOuAtualizarDragReordenando(AdmMenu menuPaiAntigo, AdmMenu menuPaiNovo, AdmMenu menuMover)
        {
            cacheMenuEstaticos.Clear();
            if ((menuPaiAntigo != null) && (!menuPaiAntigo.Equals(menuPaiNovo)))
            {
                menuPaiAntigo = atualizaSubmenu(menuPaiAntigo);
            }
            if (menuPaiNovo != null)
            {
                menuPaiNovo = atualizaSubmenu(menuPaiNovo);
            }
            menuMover = load(menuMover.Id);
            menuMover.AdmMenuPai = menuPaiNovo;
            return update(menuMover);
        }

        private AdmMenu atualizaSubmenu(AdmMenu menuPai)
        {
            if ((menuPai.AdmSubMenus != null) && (menuPai.AdmSubMenus.Count > 0))
            {
                foreach (AdmMenu menu in menuPai.AdmSubMenus)
                {
                    update(menu);
                }
            }
            return menuPai;
        }

        public AdmMenu salvarOuAtualizar(AdmMenu novoItemMenu)
        {
            cacheMenuEstaticos.Clear();
            if ((novoItemMenu.AdmMenuPai != null)) // && (novoItemMenu.AdmMenuPai.Id != null))
            {
                novoItemMenu.AdmMenuPai = load(novoItemMenu.AdmMenuPai.Id);
            }
            return update(novoItemMenu);
        }

        public void apagar(AdmMenu menu)
        {
            cacheMenuEstaticos.Clear();
            IList<AdmMenu> listaMenuFilhos = repositorio.findMenusFilhos(menu);
            if ((listaMenuFilhos != null) && (listaMenuFilhos.Count > 0))
            {
                foreach (AdmMenu menuFilho in listaMenuFilhos)
                {
                    apagar(menuFilho);
                }
            }
            menu = load(menu.Id);
            if (menu != null)
            {
                delete(menu);
            }
        }

        public List<MenuVO> toListaMenuVO(List<AdmMenu> listaMenu)
        {
            List<MenuVO> lista = new List<MenuVO>();
            foreach (AdmMenu menu in listaMenu)
            {
                lista.Add(menu.ToMenuVO());
            }
            return lista;
        }

    }
}

File: java-hfsframework-primefaces/hfsframework-admin/src/main/java/br/com/hfsframework/admin/business/AdmMenuBC.java
/**
 * <p><b>HFS Framework</b></p>
 * @author Henrique Figueiredo de Souza
 * @version 1.0
 * @since 2017
 */
package br.com.hfsframework.admin.business;

import java.util.ArrayList;
import java.util.List;

import javax.transaction.Transactional;

import br.com.hfsframework.admin.data.AdmMenuRepository;
import br.com.hfsframework.admin.model.AdmMenu;
import br.com.hfsframework.base.BaseBusinessController;
import br.com.hfsframework.security.model.MenuVO;

/**
 * The Class AdmMenuBC.
 */
public class AdmMenuBC extends BaseBusinessController<AdmMenu, Long, AdmMenuRepository> {

	/** The Constant serialVersionUID. */
	private static final long serialVersionUID = 1L;

	/** The cache menu estaticos. */
	private static List<AdmMenu> cacheMenuEstaticos = new ArrayList<AdmMenu>();

	/* (non-Javadoc)
	 * @see br.com.hfsframework.base.BaseBusinessController#getDescricao(java.io.Serializable)
	 */
	@Override
	public String getDescricao(Long id){
		return repositorio.getDescricao(id);
	}
	
	/* (non-Javadoc)
	 * @see br.com.hfsframework.base.BaseBusinessController#existeNovo(java.lang.String)
	 */
	@Override
	public boolean existeNovo(String novo){
		Long existe = repositorio.countNovo(novo.toLowerCase());
		return (existe > 0);
	}
	
	/* (non-Javadoc)
	 * @see br.com.hfsframework.base.BaseBusinessController#existeAntigoNovo(java.lang.String, java.lang.String)
	 */
	@Override
	public boolean existeAntigo(Long id, String novo){
		String antigo = getDescricao(id);
		Long existe = repositorio.countAntigo(antigo.toLowerCase(), novo.toLowerCase());
		return (existe > 0);
	}
	
	/**
	 * Salvar ou atualizar drag reordenando.
	 *
	 * @param menuPaiAntigo
	 *            the menu pai antigo
	 * @param menuPaiNovo
	 *            the menu pai novo
	 * @param menuMover
	 *            the menu mover
	 * @return the adm menu
	 */
	@Transactional
	public AdmMenu salvarOuAtualizarDragReordenando(AdmMenu menuPaiAntigo, AdmMenu menuPaiNovo, AdmMenu menuMover) {
		cacheMenuEstaticos.clear();
		if ((menuPaiAntigo != null) && (!menuPaiAntigo.equals(menuPaiNovo))) {
			menuPaiAntigo = atualizaSubmenu(menuPaiAntigo);
		}
		if (menuPaiNovo != null) {
			menuPaiNovo = atualizaSubmenu(menuPaiNovo);
		}
		menuMover = (AdmMenu) load(menuMover.getId().longValue());
		menuMover.setAdmMenuPai(menuPaiNovo);
		return (AdmMenu) update(menuMover);
	}

	/**
	 * Atualiza submenu.
	 *
	 * @param menuPai
	 *            the menu pai
	 * @return the adm menu
	 */
	@Transactional
	private AdmMenu atualizaSubmenu(AdmMenu menuPai) {
		if ((menuPai.getAdmSubMenus() != null) && (!menuPai.getAdmSubMenus().isEmpty())) {
			for (AdmMenu menu : menuPai.getAdmSubMenus()) {
				update(menu);
			}
		}
		return menuPai;
	}

	/**
	 * Salvar ou atualizar.
	 *
	 * @param novoItemMenu
	 *            the novo item menu
	 * @return the adm menu
	 */
	@Transactional
	public AdmMenu salvarOuAtualizar(AdmMenu novoItemMenu) {
		cacheMenuEstaticos.clear();
		if ((novoItemMenu.getAdmMenuPai() != null) && (novoItemMenu.getAdmMenuPai().getId() != null)) {
			novoItemMenu.setAdmMenuPai((AdmMenu) load(novoItemMenu.getAdmMenuPai().getId().longValue()));
		}
		return (AdmMenu) update(novoItemMenu);
	}

	/**
	 * Apagar.
	 *
	 * @param menu
	 *            the menu
	 */
	@Transactional
	public void apagar(AdmMenu menu) {
		cacheMenuEstaticos.clear();
		List<AdmMenu> listaMenuFilhos = repositorio.findMenusFilhos(menu);
		if ((listaMenuFilhos != null) && (!listaMenuFilhos.isEmpty())) {
			for (AdmMenu menuFilho : listaMenuFilhos) {
				apagar(menuFilho);
			}
		}
		menu = (AdmMenu) load(menu.getId().longValue());
		if (menu != null) {
			delete(menu);
		}
	}

	/**
	 * To lista menu VO.
	 *
	 * @param listaMenu
	 *            the lista menu
	 * @return the list
	 */
	public List<MenuVO> toListaMenuVO(List<AdmMenu> listaMenu){
		List<MenuVO> lista = new ArrayList<MenuVO>();
		for (AdmMenu menu : listaMenu) {
			lista.add(menu.toMenuVO());
		}		
		return lista;
	}

}
